src/Workspaces/Remote/Core/Services/Options/FileWatcher.cs

```diff
@@ -0,0 +1,127 @@
+
+namespace Microsoft.CodeAnalysis.Editor.Options
+{
+    [Export(typeof(IFileWatcher)), Shared]
+    internal sealed class FileWatcher : IFileWatcher
+    {
+        private ImmutableDictionary<(string fileName, string directoryPath), FileSystemWatcher> _watchers = ImmutableDictionary<(string fileName, string directoryPath), FileSystemWatcher>.Empty;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public FileWatcher()
+        {
+        }
+
+        public event ConventionsFileChangedAsyncEventHandler ConventionFileChanged;
+        public event ContextFileMovedAsyncEventHandler ContextFileMoved;
+
+        public void Dispose()
+        {
+            var watchers = Interlocked.Exchange(ref _watchers, null);
+            if (watchers == null)
+            {
+                return;
+            }
+
+            ConventionFileChanged = null;
+            ContextFileMoved = null;
+        }
+
+        public void StartWatching(string fileName, string directoryPath)
+        {
+            FileSystemWatcher watcher = null;
+            var updated = ImmutableInterlocked.Update(
+                ref _watchers,
+                watchers =>
+                {
+                    if (watchers is null)
+                    {
+                        watcher?.Dispose();
+                        throw new ObjectDisposedException(nameof(FileWatcher));
+                    }
+
+                    if (!watchers.TryGetValue((fileName, directoryPath), out var existingWatcher))
+                    {
+                        if (watcher == null)
+                        {
+                            watcher = new FileSystemWatcher(directoryPath, fileName);
+                        }
+
+                        return watchers.Add((fileName, directoryPath), watcher);
+                    }
+
+                    return watchers;
+                });
+
+            if (updated)
+            {
+                FileSystemEventHandler handler = (sender, e) => HandleFileSystemEvent(fileName, directoryPath, e);
+                watcher.Changed += handler;
+                watcher.Created += handler;
+                watcher.Deleted += handler;
+                watcher.Renamed += HandleRenamedEvent;
+            }
+            else
+            {
+                watcher?.Dispose();
+            }
+        }
+
+        private void HandleFileSystemEvent(string fileName, string directoryPath, FileSystemEventArgs e)
+        {
+            if (e.ChangeType.HasFlag(WatcherChangeTypes.Created))
+            {
+                ConventionFileChanged?.Invoke(this, new ConventionsFileChangeEventArgs(fileName, directoryPath, ChangeType.FileCreated));
+            }
+
+            if (e.ChangeType.HasFlag(WatcherChangeTypes.Changed))
+            {
+                ConventionFileChanged?.Invoke(this, new ConventionsFileChangeEventArgs(fileName, directoryPath, ChangeType.FileCreated));
```

Comment:
Not sure if intentional to map Changed -> Created or a typo. Combine this with the previous as an || if intended?

Reply:
Was not intentional.